### PR TITLE
Improve performance of subst by removing regexp

### PR DESCRIPTION
**Pull request by lundibundi**
Should be finished but I haven't extensively tested this, therefore, WIP for now.

As we haven't finished our benchmarking I tested on node benches.
(2 mil iterations, 30 runs each configuration)
```
                                                                                        confidence improvement accuracy (*)   (**)  (***)
 tempBench/subst.js asdf@aaa@aaa, 2000000                                                      ***     73.04 %       ±4.15% ±5.55% ±7.27%
 tempBench/subst.js df@aaa@ asdklf jas 3r ,., @bbb@@ccc@ @ccc@ asdasdasdwwwwww, 2000000        ***     55.51 %       ±3.37% ±4.49% ±5.87%
```
bench https://gist.github.com/lundibundi/b82eaca7833dd649f496497a0c64f82b

### CODE REVIEW

File: lib/strings.js [hunk 1]
```diff
@@ -51,19 +71,30 @@ const subst = (
...
     }
-    if (escapeHtml) value = htmlEscape(value);
-    return value;
-  })
-);
+    if (!escapeHtml) {
+      result += value;
+    } else {
+      result += htmlEscape(value);
+    }
```
Comment (tshemsedinov) [inline lib/strings.js]: `result += escapeHtml ? htmlEscape(value) : value;`

File: lib/strings.js [hunk 2]
```diff
@@ -26,23 +26,43 @@ const htmlEscape = (
...
-) => (
-  tpl.replace(SUBST_REGEXP, (s, key) => {
-    const pos = key.indexOf('.');
-    const name = pos === 0 ? dataPath + key : key;
-    let value = getByPath(data, name);
+) => {
+  let start = 0;
+  let end = tpl.indexOf('@');
+  if (end === -1) {
+    return tpl;
```
Comment (tshemsedinov) [inline lib/strings.js]: `if (end === -1) return tpl;`

File: lib/strings.js [hunk 3]
```diff
@@ -26,44 +26,65 @@ const htmlEscape = (
...
+
+    let value = getByPath(d, key);
     if (value === undefined) {
-      if (key === '.value') {
-        value = getByPath(data, dataPath);
-      } else {
-        value = '[undefined]';
-      }
+      if (key === '.value') value = d;
+      else value = '[undefined]';
```
Comment (tshemsedinov) [inline lib/strings.js]: `value = key === '.value' ? d '[undefined]';`

File: lib/strings.js [hunk 4]
```diff
@@ -26,44 +26,65 @@ const htmlEscape = (
...
+      start -= 1;
+      break;
+    }
+
+    let key = tpl.slice(start, end);
+    let d = data;
+    if (key.startsWith('.')) {
+      key = key.slice(1);
+      d = defaultData;
+    }
```
Comment (tshemsedinov) [inline lib/strings.js]:
```js
if (key.startsWith('.')) {
  key = key.slice(1);
  d = defaultData;
} else {
  key = tpl.slice(start, end);
  d = data;
}
```

Comment (lundibundi) [reply lib/strings.js]: Slightly different but this indeed seems better.

**Comment by lundibundi** [comment]
Final review request.

File: lib/strings.js [hunk 5]
```diff
@@ -26,44 +26,67 @@ const htmlEscape = (
...
+  if (end === -1) return tpl;
+
+  const defaultData = getByPath(data, dataPath);
+  let result = '';
+  while (end !== -1 && start < tpl.length) {
+    result = tpl.substring(start, end);
+    start = end + 1;
+    end = tpl.indexOf('@', start);
+    if (end === -1) {
+      start -= 1;
```
Comment (tshemsedinov) [inline lib/strings.js]: Why not `start--` ?

Comment (lundibundi) [reply lib/strings.js]: Oh well, yep. :smile:

**Comment by lundibundi** [comment]
Landed in d594dc988847e3fe3d6ada853d83d30a40c0f649.

**Pull request closed without merging**